Changes

nowszy94 opened the pull request:
Following Tomek's suggestions changed and fixed several bugs/issues. BookRepositoryInMemory is still not fully thread save, and I going to fix it in the future.

nowszy94 committed:
Tomek's sugestions changes

- Book fields tile and author are now final
- added BookBuilder
- deleted mapping for /books/addBasicBooks, and created PostConstruct
method in BookRepositoryInMemory
- Changed name of MyExample class to SprinBootApplication

nowszy94 committed:
Deleted target folder from repository

nowszy94 committed:
Updated controller, repository and changed data init method

Controller have more mapping now. Added /authors, /authors/{authorName}
and /titles/{titleName}. Repository have now 2 extra methods,
finidByTitle, and findByAuthor, and implemented it in InMemory
implementation. Added CommandLineRunner in SpringBootApplication class
in place of PostContruct method in BookRepositoryInMemory.

nowszy94 committed:
Lombok integration

Added lombok dependency, and modified Book class

nowszy94 committed:
Changed repository class, and now throwing Exceptions.

BookController gained method for ExceptionHandler.
Need to be test more.

nowszy94 committed:
Changed repository class, and now throwing Exceptions.

nowszy94 committed:
Package name, and exception handling

Changed main package name to com.cognifide.
Created wxception - BookNotFoundException, and edited
BookRepositoryInMemory that is throwing that exception whenever book
cannot be list or deleted.

tomaszmichalak commented on pom.xml:
```diff
@@ -21,6 +21,12 @@
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+            <version>1.16.8</version>
```
Please move version to properties section.
Additionally check who your IDE support lombok while debugging.

puradawid commented on src/main/java/com/cognifide/SpringBootApplication.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Created by cognifide.nowak on 01.04.2016.
+ */
+@org.springframework.boot.autoconfigure.SpringBootApplication
```
Probably you didn't want to use fully qualified name here (in this case `import` is welcomed). You should use fully qualified name only when the class name is ambiguous.

nowszy94 replied:
In this case the name is ambiguous, because class, that is annotated with this annotation, is also named SpringBootApplication. I'm going to change it in the future.

puradawid commented on src/main/java/com/cognifide/SpringBootApplication.java:
```diff
@@ -0,0 +1,36 @@
+    public CommandLineRunner dataInitializer(BookRepository br) {
+        return new CommandLineRunner() {
+            public void run (String... args) throws Exception {
+                // initialize bookRepository
```
Comments can get outdated pretty fast. However, this one is valueable. Maybe you can place this knowledge (that tells pupropse of instructions below) directly to code?

puradawid commented on src/main/java/com/cognifide/controller/BookController.java:
```diff
@@ -0,0 +1,63 @@
+        return bookRepository.dropAll();
+    }
+
+    @RequestMapping(value = "/title/{title}", method = RequestMethod.GET)
```
If this API was specified in advance, that's fine, but anyway it is not consistent with REST methods you did above:

`GET /books` - lists all books
`GET /books/:id` - particular book with given ID
`DELETE /books/:id` - removes book
`PUT /books/:id` - alters book

Those are totally fine and common used as a REST. However, you can't do this (from REST definition perspective):

`GET /books/title/:title`

because of its filtering of existing collection and you suggest here that have some subresource called `title`. You probably should have something for filtering using query, like:

`GET /books?title=MyTitle`

And this approach is closer to REST definition.

puradawid commented on src/main/java/com/cognifide/controller/BookController.java:
```diff
@@ -0,0 +1,63 @@
+        return authors;
+    }
+
+    @RequestMapping(value = "/author/{author}", method = RequestMethod.GET)
```
[Same rule](https://github.com/nowszy94/SpringBootTest/pull/3/files#r60218153) applies here.

puradawid commented on src/main/java/com/cognifide/controller/BookController.java:
```diff
@@ -0,0 +1,63 @@
+        return bookRepository.findByTitle(title);
+    }
+
+    @RequestMapping(value = "/author", method = RequestMethod.GET)
```
`GET /books/author` is not REST as well. You should have separate endpoint (not related to `book` resource) for `author`.

puradawid replied:
However, getting an author name of concrete book by `GET /books/:id/author` is totally fine.

puradawid commented on src/main/java/com/cognifide/repository/BookRepositoryInMemory.java:
```diff
@@ -0,0 +1,74 @@
+public class BookRepositoryInMemory implements BookRepository {
+
+    private List<Book> bookList = Collections.synchronizedList(new ArrayList<Book>());
+
```
Doubled newline

puradawid commented on src/main/java/com/cognifide/repository/BookRepositoryInMemory.java:
```diff
@@ -0,0 +1,74 @@
+
+    public List<Book> findByAuthor(String author) {
+        List<Book> results = new ArrayList<Book>();
+        for (Book book : bookList) {
```
If you are using Java 8 `filter` method is more appropriate here.

puradawid commented on src/main/java/com/cognifide/repository/BookRepositoryInMemory.java:
```diff
@@ -0,0 +1,74 @@
+            if(book.getAuthor().equals(author))
+                results.add(book);
+        }
+        if(results.size() == 0)
```
`if (booleanValue())` should have a space between `if` and parenthesis, otherwise `for` and `if` formatting are inconsistent.

puradawid commented on src/main/java/com/cognifide/repository/BookRepositoryInMemory.java:
```diff
@@ -0,0 +1,74 @@
+    public List<Book> dropAll() {
+        if(bookList.size() == 0)
+            throw new BookNotFoundException();
+        List<Book> books = new ArrayList<Book>(bookList.size());
```
No need to specify `Book` type in implementation, since Java 7 there is a `diamond operator` already, so:

`List<Book> books = new ArrayList<>(bookList.size());`

And it should work well, DRY!

nowszy94 committed:
Changed main class name to MySpringBootApplication. Removed comment in dataInitializer method, and changed that method name to more unequivocal.
In pom.xml, created property section including lombok, and spring-boot versions.

nowszy94 committed:
Add Author Controller class
Refactor

nowszy94 committed:
Add properties section in pom.xml
